### ISSUE TITLE
spawnProcess on unix doesn't handle case where os.fork fails cleanly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10979]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#1410|
|Type|defect|
|Created|2006-01-10 03:43:02Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1410 1410
type__defect defect
reporter__itamarst itamarst
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1136864582000000 1136864582000000
changetime__1212483028000000 1212483028000000
version__ 
owner__ 
cc__glyph cc__exarkun cc__itamarst cc__jknight cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Either it returns -1, or more likely Python throws exception; in either case the
error should be forwarded to user in some non-platform-specific way.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Oh Oh IProcessProtocol.connectionFailed()
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
While you're completely correct that this should be handled, I can't help but be
curious - *how* did you run across this bug?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Oh, also execv() can fail.  We should handle that too.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Coworker ran across this problem in work code he was debugging.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
Yeah.  That one (execv) seems like it would actually be a bit more common; execv
can fail if the file isn't there, if the file isn't executable, if it's mounted
on a noexec partition... fork() only fails if you're out of processes, yes?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
fork() is what co-worker caught. That's the only case I know where it'll fail,
yeah, but consider that most users have fairly low process rlimits.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Doing something better upon a failure in exec is somewhat difficult. Might want 
to check out what subprocess does. It is very clever: it opens a pipe and waits 
for a pickled exception object to be sent across it, at which point it raises 
said exception in the parent process. The FD is set close-on-exec, so once the 
actual child process is exec'd, the fd goes away, and the parent can be happy.

I'm not sure all that is really necessary, though. Right now we just log all 
failures to the subprocess' stderr which seems pretty reasonable to me.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

I may do this after [#2305](https://github.com/twisted/twisted/issues/2305). It may need a new method as exarkun said in ProcessProtocol. Also, I noticed that there is no IProcessProtocol, something that would probably be handy for documentation purpose.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

Thinking of it, adding a new method creates a big problem: currently spawnProcess fails synchronously; if we postpone errors in a connectionFailed method, but the user doesn't implement it, we have to find a place where to raise the exception. One (temporary?) solution may be to call processEnded, but I don't see the advantage of this solution against raising an exception directly.

Also, we shouldn't forget that these errors are generally critical (hitting resource limits). The best thing is to be the noisiest we can.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

I guess this is actually simple if we look at it the right way: synchronous errors can reasonably be reported synchronously; asynchronous errors cannot be.  So all we have to do is decide whether these failures are synchronous or not (or if some are one and some are the other).  ;)

Based on the current implementation, errors from fork are clearly synchronous and errors from execv are less clear.  So (again, based on the current implementation), we could raise an exception if fork fails.  For execv, if we had the parent process wait for its success or failure, then its errors could be treated synchronously and we could raise them.  This may not be desirable though - waiting for the completion of execv might block for an unreasonable amount of time (particularly if you are someone crazy who uses NFS).

So, I'd say that execv errors are actually asynchronous.  If that's the case, then we need a way to inform someone of this failure, and we can't do it by raising.

As an aside, the behavior James described of subprocess is one I've wanted in Twisted for a while (though the discussion on this ticket may actually be covering multiple separate tasks and we might want some more tickets before we're done here).  However, this is independent of whether execv failures are sync or async, since it can provide value either way.

So, if we accept that execv errors are asynchronous, then it's hardly a stretch at all to suggest that, perhaps for consistency, treating fork errors as asynchronous and feeding them to the same API makes sense.

Or we can consider another case...  Consider reactor.spawnProcess() before reactor.run().  Currently, this is buggy, and if the child exits before the SIGCHLD handler is installed, it'll get lost.  One solution to this would be to delay the actual fork/exec until the reactor is running, rather than doing it synchronous in the call to spawnProcess.  Now fork errors really are asynchronous, too.

If we do add a callback to ProcessProtocol for these, then we can supply a base implementation that logs them or something, which might satisfy the "be noisy" requirement.  Whether it is actually reasonable to add a new method to ProcessProtocol at this point, I'm not actually sure.  I _want_ it to be, but I haven't gotten much further than that.

So, ah, there's some more of my thoughts on the topic. :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

OK, I think I mostly agree with you, so let's try to advance here. Clearly adding a new method to ```ProcessProtocol``` is needed. I see the following steps:
 * create ```IProcessProtocol``` to document the current methods
 * add the method connectionFailed on the interface, implement it by default with log, and use it for some common case (remove the assert in process?).
 * manage errors in fork and pipe creation ([#2305](https://github.com/twisted/twisted/issues/2305))
 * manage errors in execv, using the subprocess trick

Each of these steps would need its own ticket (not sure about 2/3 split). What do you think?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set status to `closed`|
|-|-|

(In [23846]) Merge pipe-leak-2305-3

Author: therve
Reviewer: exarkun
Fixes [#2305](https://github.com/twisted/twisted/issues/2305)
Fixes [#10979](https://github.com/twisted/twisted/issues/10979)

Handle errors raised by os.pipe and os.fork in spawnProcess, to make sure that
opened file descriptors are closed and that UID settings are restored.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
